src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java

```diff
@@ -0,0 +1,156 @@
+
+        if (client == null) {
+            throw new ResolverException("HTTP client may not be null");
+        }
+        httpClient = client;
+
+        try {
+            metadataURI = new URI(metadataURL);
+        } catch (final URISyntaxException e) {
+            throw new ResolverException("Illegal URL syntax", e);
+        }
+    }
+
+    public String getMetadataURI() {
+        return metadataURI.toASCIIString();
+    }
+
+    @Override
+    protected void doDestroy() {
+        if (httpClient instanceof AutoCloseable) {
+            try {
+                ((AutoCloseable) httpClient).close();
+            } catch (final Exception e) {
+                log.error("Error closing HTTP client", e);
+            }
+        }
+        httpClient = null;
+        metadataURI = null;
+        cachedMetadataETag = null;
+        cachedMetadataLastModified = null;
+
+        super.doDestroy();
+    }
+
+    @Override
+    protected String getMetadataIdentifier() {
+        return metadataURI.toString();
+    }
+
+    @Override
+    protected byte[] fetchMetadata() throws ResolverException {
+        final HttpGet httpGet = buildHttpGet();
+        final HttpClientContext context = HttpClientContext.create();
+
+        try {
+            log.debug("{} Attempting to fetch metadata document from '{}'", getLogPrefix(), metadataURI);
+            return httpClient.execute(httpGet, context, response1 -> {
+                final int httpStatusCode = response1.getCode();
+                if (httpStatusCode == HttpStatus.SC_NOT_MODIFIED) {
+                    log.debug("{} Metadata document from '{}' has not changed since last retrieval", getLogPrefix(), getMetadataURI());
+                    return null;
+                }
+                if (httpStatusCode != HttpStatus.SC_OK) {
+                    final String errMsg = "Non-ok status code " + httpStatusCode + " returned from remote metadata source " + metadataURI;
+                    log.error("{} " + errMsg, getLogPrefix());
+                    throw new HttpException(errMsg);
+                }
+
+                processConditionalRetrievalHeaders(response1);
+                try {
+                    return getMetadataBytesFromResponse(response1);
+                } catch (ResolverException e) {
+                    final String errMsg = "Error retrieving metadata from " + metadataURI;
+                    throw new HttpException(errMsg, e);
+                }
+            });
+        } catch (final IOException e) {
+            final String errMsg = "Error retrieving metadata from " + metadataURI;
+            log.error("{} {}: {}", getLogPrefix(), errMsg, e.getMessage());
+            throw new ResolverException(errMsg, e);
+        }
+    }
+
+    protected HttpGet buildHttpGet() {
+        final HttpGet getMethod = new HttpGet(getMetadataURI());
+
+        if (cachedMetadataETag != null) {
+            getMethod.setHeader("If-None-Match", cachedMetadataETag);
```

Comment:
I think better to use per-defined constants for headers. e.g. `org.apache.http.HttpHeaders`